src/logging.rs

```diff
@@ -0,0 +1,237 @@
+
+    let (discord_layer_reloadable, log_reload_handle) = reload::Layer::new(discord_layer);
+
+    let discord_layer_filtered = discord_layer_reloadable
+        .with_filter(filter::Targets::new().with_target("discord", Level::INFO));
+
+    tracing_subscriber::registry()
+        .with(discord_layer_filtered)
+        .with(fmt::layer().compact())
+        .with(fmt::layer().compact().with_writer(rolling_file_writer))
+        .init();
+
+    info!("Setup logging");
+
+    let _ = DISCORD_LAYER_CHANGE_HANDLE.set(log_reload_handle);
+
+    worker_guard
+}
+
+/// Panics if called twice
+pub async fn setup_discord_logging(framework: Arc<bot::Framework>, db: Pool<MySql>) {
+    modify_discord_layer(|discord_layer| {
+        discord_layer.poise_framework = Some(framework.clone());
+    });
+
+    let http = &framework.client().cache_and_http.http;
+
+    // Setup main logging guild/channel
+
+    let main_guild_id: u64 = env::var("MAIN_GUILD_ID")
+        .expect("No main logging guild id configured")
+        .parse()
+        .expect("Could not parse MAIN_GUILD_ID env variable");
+
+    let main_guild = http
+        .get_channels(main_guild_id)
+        .await
+        .expect("Could not get main guild");
+
+    let main_logging_channel = main_guild[0].id;
+
+    modify_discord_layer(|discord_layer| {
+        discord_layer.main_log_channel = NonZeroU64::new(main_logging_channel.0);
+    });
+
+    // Setup logging channels per server
+    let guild_to_log_channel = mysql_lib::get_all_guilds(&db).await.into_iter()
+        .filter_map(|guild| {
+            guild.logger_pipe_channel.map(|logger_pipe_channel|
+                (guild.guild_id.0, logger_pipe_channel.0))
+        })
+        .collect();
+
+    modify_discord_layer(|discord_layer| {
+        discord_layer.guild_to_log_channel = guild_to_log_channel;
+    });
+}
+
+fn modify_discord_layer(f: impl FnOnce(&mut DiscordTracingLayer)) {
+    let result = DISCORD_LAYER_CHANGE_HANDLE.get().unwrap().modify(f);
+
+    if let Err(err) = result {
+        error!(
+            error = err.to_string(),
+            "Something went wrong while trying to modify the discord tracing layer"
+        );
+    }
+}
+
+#[allow(dead_code)]
+/// Panics if called before [`install_framework`]
+pub fn add_per_server_logging(guild_id: u64, log_channel_id: u64) {
+    let layer_change_handle = DISCORD_LAYER_CHANGE_HANDLE.get().unwrap();
+    let result = layer_change_handle.modify(|layer| {
+        layer.guild_to_log_channel.insert(guild_id, log_channel_id);
+    });
+    if let Err(err) = result {
+        error!(
+            error = err.to_string(),
+            "Failed to install poise framework into discord tracing layer"
+        );
+    }
+}
+
+struct DiscordTracingLayer {
+    main_log_channel: Option<NonZeroU64>,
+    poise_framework: Option<Arc<bot::Framework>>,
+    /// HashMap of GuilId's -> ChannelId's
+    guild_to_log_channel: HashMap<u64, u64>,
```

Comment:
Is there a big reason why you are using `NonZeroU64`  and `u64` instead of `ChannelId` and `GuildId`?

Reply:
I use NonZeroU64 because I think it's more expressive than u64, in that it signals that the main log channel can be missing. No reason for the u64 though, I guess in the HashMap it would be a bit better to use GuildId's and ChannelId's

Reply:
>  it signals that the main log channel can be missing

isn't that also the case with `Option<ChannelId>`?

Reply:
It is, but that leaves open the possibility of the ChannelId being zero even though the Option is Some, and uses more memory (since the size of `Option<u64>` > size of `Option<NonZeroU64>`). Does it make a difference? Not really, because ChannelId's and u64's are practically the same except for Deserialization/Serialization, but I like it this way. If you want it differently you can change it, I don't mind

Reply:
Yeah ok, then it is fine, leave it as is